interpreter/cling/lib/Interpreter/IncrementalJIT.cpp

```diff
@@ -388,6 +388,21 @@ void IncrementalJIT::addModule(const std::shared_ptr<llvm::Module>& module) {
   // default.
   module->setDataLayout(m_TMDataLayout);
 
+  // Reset the sections of all functions so that they end up in the same text
+  // section. This is important for TCling on macOS to catch exceptions raised
+  // by constructors, which requires unwinding information. The addresses in
+  // the __eh_frame table are relocated against a single __text section when
+  // loading the MachO binary, which breaks if the call sites of constructors
+  // end up in a separate init section.
+  // (see clang::TargetInfo::getStaticInitSectionSpecifier())
```

Comment:
Looks like the orcv1 (which cling is currently based on) is frozen. Even for bugs of that kind. The [orcv2 EH support](https://github.com/llvm/llvm-project/blob/main/llvm/lib/ExecutionEngine/JITLink/EHFrameSupport.cpp) seems superior and it does not seem trivial to extend the [RuntimeDyldMachOCRTPBase<Impl>::processFDE](https://github.com/llvm/llvm-project/blob/810572cc96e99b5d6fb695f1f047c2ae6507831b/llvm/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldMachO.cpp#L275) ourselves.

If that workaround is here to stay, please add a cling-standalone test with that commit. This will help us remove it when switching to orcv2 (requires upgrading to llvm12).

Reply:
> Looks like the orcv1 (which cling is currently based on) is frozen. Even for bugs of that kind.

OrcV1 is entirely removed in `master`: https://github.com/llvm/llvm-project/commit/6154c4115cd4b78d0171892aac21e340e72e32bd

> The orcv2 EH support seems superior and it does not seem trivial to extend the RuntimeDyldMachOCRTPBase::processFDE ourselves. [...] This will help us remove it when switching to orcv2 (requires upgrading to llvm12).

FWIW I don't think that using OrcV2 helps in any way: It's still relying on `RuntimeDyld` including the MachO relocation logic. And I don't see the issue with multiple text sections fixed in `master`.

> If that workaround is here to stay, please add a cling-standalone test with that commit.

Will do. I had already started with this yesterday, but running the tests from a build of ROOT seems very hard...

Reply:
> running the tests from a build of ROOT seems very hard...

I can help with that! `-Dclingtest=On` - and this patch:
```diff
diff --git a/interpreter/cling/test/lit.cfg b/interpreter/cling/test/lit.cfg
index c9441bb30f..df72976a72 100644
--- a/interpreter/cling/test/lit.cfg
+++ b/interpreter/cling/test/lit.cfg
@@ -317,23 +317,6 @@ if platform.system() not in ['Windows'] or lit_config.getBashPath() != '':
 if platform.system() not in ['Windows']:
     config.available_features.add('not_system-windows')
 
-# Do we have cling and clang sources under llvm? Some tests
-# require it.
-if not os.path.exists(os.path.join(config.test_exec_root, "..", "..", "clang", "test")):
-  config.test_exec_root = os.path.join(config.test_exec_root, "..", "..")
-  llvm_revision = urlopen(
-      "https://raw.githubusercontent.com/root-project/cling/master/LastKnownGoodLLVMSVNRevision.txt"
-      ).readline().strip().decode('utf-8')
-  subprocess.Popen(['sudo svn export http://llvm.org/svn/llvm-project/llvm/branches/{0}/utils/lit utils/lit'.format(llvm_revision)],
-                                    cwd=config.llvm_src_root,
-                                    shell=True,
-                                    stdin=subprocess.PIPE,
-                                    stdout=None,
-                                    stderr=subprocess.STDOUT).communicate('yes'.encode('utf-8'))
-config.substitutions.append(('%testexecdir', config.test_exec_root))
-config.substitutions.append(('%llvmsrcdir', config.llvm_src_root))
-config.available_features.add('vanilla-cling')
-
 if lit.util.which('libcudart.so', config.environment.get('LD_LIBRARY_PATH','')) is not None:
   config.available_features.add('cuda-runtime')
 
```

Reply:
> I can help with that! `-Dclingtest=On` - and this patch:

Thanks for that option, probably saved me a lot - I only passed `-DCLING_INCLUDE_TESTS=ON` (why are there two?!)

Reply:
Done. I tested that the added checks would have failed before the fix.